theano/tensor/basic.py

```diff
@@ -6144,9 +6143,18 @@ def perform(self, node, inputs, outputs):
             z[0] = z[0].copy()
 
     def grad(self, inputs, gout):
+        """
+        The following code is moved from tensor.nlinalg.ExtractDiag, only works
+        for matrices.
+        """
+        warnings.warn("gradient of theano.tensor.nlinalg.ExtractDiag only"
+                      "works for matrices.")
```

Comment:
Don't print a warning all the time.  Instead error out when not dealing with a matrix.